subprojects/core/src/integTest/groovy/org/gradle/api/services/BuildServiceIntegrationTest.groovy

```diff
@@ -72,33 +84,202 @@ class BuildServiceIntegrationTest extends AbstractIntegrationSpec {
         succeeds 'broken'
     }
 
-    private taskUsingUndeclaredService() {
+    def "does not nag when service is used by task with an explicit usesService call and feature preview is enabled"() {
+        given:
+        serviceImplementation()
+        customTaskUsingServiceViaProperty()
         buildFile """
-            def provider = gradle.sharedServices.registerIfAbsent("counter", CountingService) {
+            def serviceProvider = gradle.sharedServices.registerIfAbsent("counter", CountingService) {
                 parameters.initial = 42
             }
 
-            tasks.register("broken") {
+            tasks.register("explicit") {
+                it.usesService(serviceProvider)
                 doFirst {
-                    provider.get().increment()
+                    serviceProvider.get().increment()
                 }
             }
         """
+        enableStableConfigurationCache()
+
+        when:
+        succeeds 'explicit'
+
+        then:
+        outputDoesNotContain "'Task#usesService'"
     }
 
-    def "service is created once per build on first use and stopped at the end of the build"() {
+    def "does not nag when service is annotated with @ServiceReference (unnamed) and feature preview is enabled"() {
+        given:
+        serviceImplementation()
+        customTaskUsingServiceViaProperty("@${ServiceReference.class.name}()")
+        buildFile """
+            def provider = gradle.sharedServices.registerIfAbsent("counterService", CountingService) {
+                parameters.initial = 10
+                maxParallelUsages = 1
+            }
+
+            task implicit(type: Consumer) {
+                counter.convention(provider)
+            }
+        """
+        enableStableConfigurationCache()
+
+        when:
+        succeeds 'implicit'
+
+        then:
+        outputDoesNotContain "'Task#usesService'"
+    }
+
+    def "can inject shared build service by name when reference is annotated with @ServiceReference('...')"() {
+        given:
+        serviceImplementation()
+        customTaskUsingServiceViaProperty("@${ServiceReference.class.name}('counter')")
+        buildFile """
+            gradle.sharedServices.registerIfAbsent("counter", CountingService) {
+                parameters.initial = 10
+                maxParallelUsages = 1
+            }
+
+            task named(type: Consumer) {
+                // reference will be set by name
+                doLast {
+                    assert !requiredServices.requiredServices.isEmpty()
```

Comment:
🟡 Can we make the assertion a little more specific here?

Reply:
```
assert requiredServices.requiredServices.any { service ->
        service.type == CountingService
}
```